mqtt/mqtt-edgehub/src/connection/delivery.rs

```diff
@@ -0,0 +1,137 @@
+use std::{collections::HashMap, sync::Arc};
+
+use async_trait::async_trait;
+use lazy_static::lazy_static;
+use mqtt_broker::{
+    BrokerHandle, Error, IncomingPacketProcessor, Message, OutgoingPacketProcessor, PacketAction,
+    SystemEvent,
+};
+use parking_lot::Mutex;
+use regex::Regex;
+
+use mqtt3::proto::{self, Packet, PacketIdentifier, Publication};
+
+pub struct PublicationDelivery<P> {
+    inner: P,
+    broker_handle: BrokerHandle,
+    waited_to_be_acked: Arc<Mutex<HashMap<PacketIdentifier, String>>>,
+}
+
+/// MQTT packet processor wrapper. It identifies `IoTHub` M2M outgoing publishes,
+/// saves PACKETID to send a confirmation packet back to sender on special topic
+/// "$edgehub/delivery".
+impl<P> PublicationDelivery<P> {
+    pub fn new(
+        broker_handle: BrokerHandle,
+        inner: P,
+        waited_to_be_acked: Arc<Mutex<HashMap<PacketIdentifier, String>>>,
+    ) -> Self {
+        Self {
+            inner,
+            broker_handle,
+            waited_to_be_acked,
+        }
+    }
+
+    fn prepare_confirmation(
+        &mut self,
+        packet: &Packet,
+    ) -> Result<Option<Publication>, serde_json::Error> {
+        const DELIVERY_TOPIC: &str = "$edgehub/delivered";
+
+        let packet_identifier = match &packet {
+            Packet::PubAck(puback) => puback.packet_identifier,
+            Packet::PubRel(pubrel) => pubrel.packet_identifier,
+            _ => return Ok(None),
+        };
+
+        let publication = match self.waited_to_be_acked.lock().remove(&packet_identifier) {
+            Some(topic_name) => Some(Publication {
+                topic_name: DELIVERY_TOPIC.into(),
+                qos: proto::QoS::AtLeastOnce,
+                retain: false,
+                payload: serde_json::to_string(&topic_name)?.into(),
+            }),
+            None => None,
+        };
+
+        Ok(publication)
+    }
+
+    fn store_packet_info(&mut self, packet: &Packet) {
+        if let Some((packet_identifier, topic_name)) = match_m2m_publish(packet) {
+            self.waited_to_be_acked
+                .lock()
+                .insert(packet_identifier, topic_name);
+        }
+    }
+}
+
+fn match_m2m_publish(packet: &Packet) -> Option<(proto::PacketIdentifier, String)> {
+    const DEVICE_ID: &str = r"(?P<device_id>[^/]+)";
+    const MODULE_ID: &str = r"(?P<module_id>[^/]+)";
+    lazy_static! {
+        static ref M2M_PUBLISH_PATTERN: Regex = Regex::new(&format!(
```

Comment:
maybe simplify regex a bit? you don't need named groups, you just need to check it if matches.